internal/nbconn/nbconn_real_non_block_windows.go

```diff
@@ -0,0 +1,149 @@
+//go:build windows
+
+package nbconn
+
+import (
+	"errors"
+	"golang.org/x/sys/windows"
+	"io"
+	"syscall"
+	"unsafe"
+)
+
+var dll = syscall.MustLoadDLL("ws2_32.dll")
+
+// int ioctlsocket(
+//
+//	[in]      SOCKET s,
+//	[in]      long   cmd,
+//	[in, out] u_long *argp
+//
+// );
+var ioctlsocket = dll.MustFindProc("ioctlsocket")
+
+type sockMode int
+
+const (
+	FIONBIO             int      = 0x8004667e
+	sockModeBlocking    sockMode = 0
+	sockModeNonBlocking sockMode = 1
+)
+
+func setSockMode(fd uintptr, mode sockMode) error {
+	res, _, err := ioctlsocket.Call(fd, uintptr(FIONBIO), uintptr(unsafe.Pointer(&mode)))
+	// Upon successful completion, the ioctlsocket returns zero.
+	if res != 0 && err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// realNonblockingWrite does a non-blocking write. readFlushLock must already be held.
+func (c *NetConn) realNonblockingWrite(b []byte) (n int, err error) {
+	if c.nonblockWriteFunc == nil {
+		c.nonblockWriteFunc = func(fd uintptr) (done bool) {
+			// Make sock non-blocking
+			if err := setSockMode(fd, sockModeNonBlocking); err != nil {
```

Comment:
I don't have a test case so feel free to ignore this comment. 
What if another goroutine ends a write and set the socket to blocking mode after this method ends and before calling `WSASend`?

Reply:
> @drakkan Connection isn't considered as thread (goroutine) safe , it should be managed through connection pool, that guarantees that only one goroutine can work with connection.
> 
> Here it is - https://github.com/jackc/pgx/wiki/Getting-started-with-pgx#using-a-connection-pool

Ok thanks. I was just trying to figure out if it could be related to #1552, #1548

Reply:
@drakkan hmm, #1552 looks really strange. I started debugging this case and found that there is inconsistent `setSockMode` calls:
```go
var currSockMode = sockModeBlocking

func setSockMode(fd uintptr, mode sockMode) error {
	if mode == currSockMode {
		fmt.Printf("double setSockMode=%d\n", mode)
	}

	currSockMode = mode

	res, _, err := ioctlsocket.Call(fd, uintptr(FIONBIO), uintptr(unsafe.Pointer(&mode)))
	// Upon successful completion, the ioctlsocket returns zero.
	if res != 0 && err != nil {
		return err
	}

	return nil
}
```

Here is output while using CopyFrom with large amount of data:
```
double setSockMode=1
double setSockMode=0
double setSockMode=1
double setSockMode=0
double setSockMode=1
double setSockMode=0
double setSockMode=0
double setSockMode=0
double setSockMode=0
double setSockMode=0
double setSockMode=1
double setSockMode=0
double setSockMode=0
double setSockMode=0
double setSockMode=0
double setSockMode=1
(hangs forever)
```

Reply:
Added goroutine ids to output:
```
double setSockMode=1 (id=1)
double setSockMode=0 (id=1)
double setSockMode=0 (id=25)
double setSockMode=0 (id=1)
double setSockMode=0 (id=25)
double setSockMode=1 (id=25)
```

Looks like there is race condition

Reply:
Just commented out these lines of code:
```go
// Make sock blocking again
if err := setSockMode(fd, sockModeBlocking); err != nil {
	c.nonblockWriteErr = err
	return true
}
```

And CopyFrom completed successfully

Reply:
@drakkan I see two possible solutions:
1. Make socket to be always in non-blocking mode (but I'm not sure how it can affect other operations)
2. Wrap Windows non-blocking read/write into critical section, e.g.:
```go
var guard = make(chan struct{}, 1)

func (c *NetConn) realNonblockingWrite(b []byte) (n int, err error) {
	if c.nonblockWriteFunc == nil {
		c.nonblockWriteFunc = func(fd uintptr) (done bool) {
			guard <- struct{}{}
// .....

	<-guard
        return true
```

Reply:
@jackc do you have any ideas?

Reply:
> Added goroutine ids to output:
> 
> ```
> double setSockMode=1 (id=1)
> double setSockMode=0 (id=1)
> double setSockMode=0 (id=25)
> double setSockMode=0 (id=1)
> double setSockMode=0 (id=25)
> double setSockMode=1 (id=25)
> ```
> 
> Looks like there is race condition

@codercms  Thanks for confirming. I did a test with this change

```
diff --git a/internal/nbconn/nbconn_real_non_block_windows.go b/internal/nbconn/nbconn_real_non_block_windows.go
index ed1662e0..de63c244 100644
--- a/internal/nbconn/nbconn_real_non_block_windows.go
+++ b/internal/nbconn/nbconn_real_non_block_windows.go
@@ -6,6 +6,7 @@ import (
        "errors"
        "golang.org/x/sys/windows"
        "io"
+       "sync"
        "syscall"
        "unsafe"
 )
@@ -21,6 +22,10 @@ var dll = syscall.MustLoadDLL("ws2_32.dll")
 // );
 var ioctlsocket = dll.MustFindProc("ioctlsocket")
 
+var (
+       mu sync.Mutex
+)
+
 type sockMode int
 
 const (
@@ -41,6 +46,9 @@ func setSockMode(fd uintptr, mode sockMode) error {
 
 // realNonblockingWrite does a non-blocking write. readFlushLock must already be held.
 func (c *NetConn) realNonblockingWrite(b []byte) (n int, err error) {
+       mu.Lock()
+       defer mu.Unlock()
+
        if c.nonblockWriteFunc == nil {
                c.nonblockWriteFunc = func(fd uintptr) (done bool) {
                        // Make sock non-blocking
@@ -92,6 +100,9 @@ func (c *NetConn) realNonblockingWrite(b []byte) (n int, err error) {
 }
 
 func (c *NetConn) realNonblockingRead(b []byte) (n int, err error) {
+       mu.Lock()
+       defer mu.Unlock()
+
        if c.nonblockReadFunc == nil {
                c.nonblockReadFunc = func(fd uintptr) (done bool) {
                        // Make sock non-blocking
```

If I run the test cases I have this

```
Z:\pgx>go test -v --count 1 -timeout 30s -run ^TestInternalNonBlockingWrite$ github.com/jackc/pgx/v5/internal/nbconn
=== RUN   TestInternalNonBlockingWrite
=== RUN   TestInternalNonBlockingWrite/Pipe
=== RUN   TestInternalNonBlockingWrite/TCP_with_Fake_Non-blocking_IO
=== RUN   TestInternalNonBlockingWrite/TLS_over_TCP_with_Fake_Non-blocking_IO
    nbconn_test.go:303:
                Error Trace:    Z:\pgx\internal\nbconn\nbconn_test.go:303
                                                        Z:\pgx\internal\nbconn\nbconn_test.go:143
                Error:          Received unexpected error:
                                local error: tls: bad record MAC
                Test:           TestInternalNonBlockingWrite/TLS_over_TCP_with_Fake_Non-blocking_IO
=== RUN   TestInternalNonBlockingWrite/TCP_with_Real_Non-blocking_IO
=== RUN   TestInternalNonBlockingWrite/TLS_over_TCP_with_Real_Non-blocking_IO
--- FAIL: TestInternalNonBlockingWrite (0.40s)
    --- PASS: TestInternalNonBlockingWrite/Pipe (0.17s)
    --- PASS: TestInternalNonBlockingWrite/TCP_with_Fake_Non-blocking_IO (0.03s)
    --- FAIL: TestInternalNonBlockingWrite/TLS_over_TCP_with_Fake_Non-blocking_IO (0.16s)
    --- PASS: TestInternalNonBlockingWrite/TCP_with_Real_Non-blocking_IO (0.02s)
    --- PASS: TestInternalNonBlockingWrite/TLS_over_TCP_with_Real_Non-blocking_IO (0.02s)
FAIL
FAIL    github.com/jackc/pgx/v5/internal/nbconn 0.939s
FAIL
```

so also the bad record MAC seems fixed for the `TLS_over_TCP_with_Real_Non-blocking_IO` test (fake non blocking IO still fails). @jackc what do you think about? Thanks

Reply:
@drakkan @jackc I found this article - https://scarff.id.au/blog/2022/go-concurrency-idioms-dont-work-for-duplex-sockets/ about Go non-blocking I/O, and it seems like concurrent read/write from net.Conn is not safe.
Then I found this package - https://github.com/lesismal/nbio (Windows is not supported)

See implementation details:
* Read - https://github.com/lesismal/nbio/blob/master/conn_unix.go#L63
* Write - https://github.com/lesismal/nbio/blob/master/conn_unix.go#L146

The both read and write operations using shared mutex to prevent concurrent read & write.

What do you think?  
Should be concurrency limited in PGX for UNIX systems too? (like one of possible solutions for Windows)

Reply:
> @drakkan @jackc I found this article - https://scarff.id.au/blog/2022/go-concurrency-idioms-dont-work-for-duplex-sockets/ about Go non-blocking I/O, and it seems like concurrent read/write from net.Conn is not safe. Then I found this package - https://github.com/lesismal/nbio (Windows is not supported)
> 
> See implementation details:
> 
> * Read - https://github.com/lesismal/nbio/blob/master/conn_unix.go#L63
> * Write - https://github.com/lesismal/nbio/blob/master/conn_unix.go#L146
> 
> The both read and write operations using shared mutex to prevent concurrent read & write.
> 
> What do you think? Should be concurrency limited in PGX for UNIX systems too? (like one of possible solutions for Windows)

@codercms on Linux if I print the goroutine id I have a different id for reads and writes. I assumed the same for Windows and since on Windows we set the blocking mode in every read and write this will explain the reported issues (at least the hangs). This is the reason I asked [this](https://github.com/jackc/pgx/pull/1525#discussion_r1139224288).
I'm not familiar enough with pgx code to understand if reading and writing from different goroutines can cause issues also on Linux/Unix

Reply:
I don't have much expertise with Windows network programming, but a bit of doc and Google searching seems to indicate that it *should* be safe to for one thread to read and one thread to write from the safe socket. However, it appears that some of the winsock API is not thread safe. Perhaps that includes the `ioctlsocket` function.

...

But now that I think about it a bit more I think I have a more likely cause for the deadlock. pgx will read and write concurrently under certain conditions (such as `bufferNonblockingRead` when a write is blocked). Both of these functions set non-blocking mode, read or write, and set blocking mode.

What happens if the following sequence of events occurs:

1. Writer sets non-blocking mode
2. Reader sets non-blocking mode
3. Writer writes
4. Writer sets blocking mode
5. Reader reads

The socket is unexpectedly in blocking mode for the read. If nothing is ever received it hangs.

---

As far as Unix goes, I think that since we are not changing the socket options, we are only reading and writing, it is safe.

---

I **really** wish the Go stdlib included non-blocking I/O. This system is so complicated...

I'm half tempted to try using goroutines and normal Go I/O again. I tried it a few years ago and the performance hit was significant, but maybe modern Go has higher performance...

Reply:
@jackc You're absolutely right about that socket can be used for concurrent read/write (socket is full duplex).
I think implementing critical section (e.g. mutex) for Windows non-blocking I/O will be simplest solution. What do you think will this fix the problem? But additional benchmark is needed, because I/O actually will be not full duplex (you can only write or only read at the same time).

Reply:
That would be the simplest approach.

But it seems it should be possible to switch to non-blocking mode at the beginning of a sequence of read or writes and turn it off at the end. Not sure how messy that would be though.

Reply:
@jackc switching to non-blocking mode looks good too, because it doesn't hit performance.
I think non-blocking conn could expose `setBlocking(blocking bool)` method or something like this, for Windows it will call `ioctlsocket`, for Unix it will do nothing.

Reply:
@jackc I made draft PR for setting connection non-blocking mode explicitly - https://github.com/jackc/pgx/pull/1555 . Please have a look on it